mutation_reader.cc

```diff
@@ -194,12 +198,25 @@ class mutation_reader_merger {
     // end, a call to next_partition() or a call to
     // fast_forward_to(dht::partition_range).
     reader_and_last_fragment_kind _single_reader;
+    // Holds a reference to the reader that previously contributed a fragment.
+    // When hits reader contributes a certain number of fragments in a row,
+    // gallop mode becomes enabled. In this mode, the _galloping_reader
+    // is assumed to be the contributor of the next fragment.
+    reader_and_last_fragment_kind _galloping_reader;
```

Comment:
I found the old discussion ([here](https://groups.google.com/forum/#!searchin/scylladb-dev/%22combined_reader$3A$20optimise$20for%22|sort:date/scylladb-dev/YtDC7dm1l0c/osVDTeevAQAJ)). The argument against removing the `_single_reader` variable is that re-using `_next` makes it less clear that the single reader mode is being turned on.

I think I agree with Paweł on that. Furthermore, we still need an additional field for tracking which reader previously contributed a fragment in order to know when to increase/clear `_gallop_mode_hits`.

Reply:
If we do what we suggested, we don't need `_gallop_mode_hits`. Gallop mode is implicit and is triggered any time a single reader is in `_next`.

Reply:
I'm not sure if I understand correctly. This optimization is supposed to kick in when one reader contributes 3 fragments consecutively, so we need to count it somehow.

`_next` holds all readers that contributed a fragment in previous `operator()()` call. `_next.size()` would be 1 also in the case when two readers contributing alternating fragments are merged - in that case galloping would be detrimental to performance, because we do an extra key comparison before falling back to heap operations.

I think it is possible to remove `_galloping_reader` and use the previous value from `_next` to compare if currently contributing reader is the same as before. We would have to pass this information to `operator()()` or structure the function differently.

Reply:
Ok, so the reason we want to trigger this optimization after a certain threshold is to avoid the additional key comparison. I think we can do this with my proposed solution as well, albeit with some changes.

Before I go into that I'd like to lay out my reasons why I'd like to do away with the separate states (single reader, multi reader and gallop). Yes, adding another member is not that terrible (like Pawel put it), yet this creates another separate logical stream within the reader. We already have two and are about to add a third one, each with its own way of fetching fragments from the reader, then propagating them downstream. It's almost like having 3 readers in one, with a bunch of duplicate code. This makes one of our most complex reader even more complex. So I think we should also weight code complexity, not just runtime performance.

I think we can make this even simpler then I originally proposed: modify `prepare_next()` to not push all new fragments into the `_fragment_heap`, instead these new fragments should go into another (unordered) container, `_new_fragments`. Then, in `operator()()` we can see how many new fragments were produced and if there is just one, make a "bet" whether we'll need to merge or not. This can be based on a history of recent such "bets". This history is `_galloping_reader_hits` in your proposal.
We start out by assuming we need to merge. We push `_new_fragments.front()` to `_fragment_heap`. If after making the heap it is still the last element (it is the smallest in the heap) we bump the counter as we initiated a merge when we didn't need one. After `N` (3) such events we flip the assumption and enter gallop mode, assuming we won't need a merge and check `_new_fragments.front()` against `_fragment_heap` before pushing it.